src/librbd/api/Mirror.cc

```diff
@@ -756,6 +757,188 @@ void Mirror<I>::image_demote(I *ictx, Context *on_finish) {
+
+  librbd::ImageCtx *remote_ictx = new librbd::ImageCtx(ictx->name.c_str(), "",
+                                                       nullptr,
+                                                       remote_io_ctx, false);
+  remote_ictx->read_only_mask &= ~IMAGE_READ_ONLY_FLAG_NON_PRIMARY;
+  r = remote_ictx->state->open(0);
+  if (r < 0) {
+    lderr(cct) << "error opening image: " << ictx->name.c_str() << ": "
+               << cpp_strerror(r) << dendl;
+    return r;
+  }
+  uint64_t remote_size;
+  {
+    std::shared_lock image_locker{remote_ictx->image_lock};
+    remote_size =  remote_ictx->get_image_size(CEPH_NOSNAP);
+  }
+  ldout(cct, 20) << "remote image size: " << remote_size  << dendl;
+  uint64_t remote_object_count = Striper::get_num_objects(remote_ictx->layout,
+                                                          remote_size);
+  ldout(cct, 20) << "remote image object count: "
+                 << remote_object_count << dendl;
+
+  bufferlist init_value_bl;
+  bufferlist csum_bl;
+  bufferlist rcsum_bl;
+  bool local_obj_absent;
+  bool remote_obj_absent;
+
+  ceph::encode(static_cast<uint32_t>(-1), init_value_bl);
+  for (uint64_t i = 0; i < object_count; ++i) {
+    csum_bl.clear();
+    local_obj_absent = false;
+    remote_obj_absent = false;
+
+    ldout(cct, 20) << "obj name: " << ictx->get_object_name(i) << dendl;
+    r = io_ctx.checksum(ictx->get_object_name(i), LIBRADOS_CHECKSUM_TYPE_CRC32C,
+                        init_value_bl, ictx->get_object_size(), 0, 0, &csum_bl);
+    if (r == -ENOENT) {
+      local_obj_absent = true;
+    } else if (r < 0) {
+      lderr(cct) << "failed to checksum object: " << cpp_strerror(r) << dendl;
+      return r;
+    }
+
+    rcsum_bl.clear();
+    ldout(cct, 20) << "remote obj name: "
+                   << remote_ictx->get_object_name(i) << dendl;
+    r = remote_io_ctx.checksum(remote_ictx->get_object_name(i),
+                              LIBRADOS_CHECKSUM_TYPE_CRC32C,
+                              init_value_bl,
+                              remote_ictx->get_object_size(),
+                              0, 0, &rcsum_bl);
+    if (r == -ENOENT) {
+      remote_obj_absent = true;
+    } else if (r < 0) {
+      lderr(cct) << "failed to checksum remote object: "
+                 << cpp_strerror(r) << dendl;
+      return r;
+    }
+
+    if (local_obj_absent && remote_obj_absent) {
+      continue;
+    } else if (local_obj_absent && !remote_obj_absent) {
+      lderr(cct) << "local image doesn't have object: "
+                 << ictx->get_object_name(i) << dendl;
+      return -ENOENT;
+    } else if (!local_obj_absent && remote_obj_absent) {
+      lderr(cct) << "remote image doesn't have object: "
+                 << ictx->get_object_name(i) << dendl;
+      return -ENOENT;
+    }
+
+    if (!csum_bl.contents_equal(rcsum_bl)) {
+      lderr(cct) << "checksum mismatch for object: "
+                 << ictx->get_object_name(i) << dendl;
+      return -EIO;
```

Comment:
Would there be any value in continuing to check the remaining objects to see all the ones that mismatch?

Reply:
Doesn't matter, we will have to resync the whole image even if one mismatch.

Reply:
Agreed on the resync. I thought it may be useful for troubleshooting (a diff between images..), but now I think more it may be outside of the scope of this issue.